### PR TITLE
feat(api): add `read` top level function for interactive analysis with CSV and Parquet files

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds a `read` top level API function for reading the following kinds of things:

1. Local CSV, TSV, TXT and Parquet files. CSV/TSV/TXT files can be gzip compressed
1. Remote versions of any of those
1. Globs of 1., mixing compressed/uncompressed is allowed.

### CODE REVIEW

**Review by jcrist — COMMENTED** [review]
This seems like a nice win here.

It also removes a wart with the current top-level `connect` where `ibis.connect` can return either a connection _or_ a table dependent on the value of the object passed in (not the type). This makes type annotating `ibis.connect` impossible to do in a nice way.

Perhaps we should deprecate loading parquet/csv files in `ibis.connect` in favor of these functions?

**Comment by jcmkk3** [comment]
I'm really happy to see these functions. I would use these commonly for how I work with Ibis, plus it makes it so much easier for folks to try out Ibis for the first time.

I can't tell from the duckdb documentation, but can these functions read from a URL? I often use [tidytuesday](https://github.com/rfordatascience/tidytuesday) datasets when I want to try something out and it is nice in pandas or dplyr to be able to read directly from the URL.

**Comment by cpcloud** [comment]
@jcmkk3 Really appreciate the rapid feedback!

I've added support for URLs in the latest commit 🎉:

```
In [1]: import ibis; ibis.options.interactive = True; from ibis import _

In [2]: t = ibis.read_csv('https://raw.githubusercontent.com/ArthurCheib/analytical-politics-project/main/dat
   ...: a/tidy-data-fh-un.csv')

In [3]: t
Out[3]:
┏━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━┓
┃ country     ┃ year  ┃ CL    ┃ PR    ┃ Status ┃ Region_Code ┃ Region_Name ┃ is_ldc ┃
┡━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━┩
│ string      │ int32 │ int32 │ int32 │ string │ int32       │ string      │ int32  │
├─────────────┼───────┼───────┼───────┼────────┼─────────────┼─────────────┼────────┤
│ Afghanistan │  1995 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  1996 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  1997 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  1998 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  1999 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2000 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2001 │     7 │     7 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2002 │     6 │     6 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2003 │     6 │     6 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2004 │     6 │     5 │ NF     │         142 │ Asia        │      1 │
│ Afghanistan │  2005 │     5 │     5 │ PF     │         142 │ Asia        │      1 │
│ …           │     … │     … │     … │ …      │           … │ …           │      … │
└─────────────┴───────┴───────┴───────┴────────┴─────────────┴─────────────┴────────┘


In [4]: t.Region_Name.value_counts()
Out[4]:
┏━━━━━━━━━━━━━┳━━━━━━━┓
┃ Region_Name ┃ count ┃
┡━━━━━━━━━━━━━╇━━━━━━━┩
│ string      │ int64 │
├─────────────┼───────┤
│ Asia        │  1218 │
│ Europe      │  1099 │
│ Africa      │  1388 │
│ Americas    │   910 │
│ Oceania     │   364 │
└─────────────┴───────┘
```

**Comment by ogrisel** [comment]
Nice feature, especially with the https URL support.

Maybe you should make it explicit in the docstring that this will read the full data in a in-memory (duckdb) table?

If the file is large, it might be recommended instead to create an empty in-memory duckdb connection and subsequently register the parquet (or CSV) file as a foreign table:

https://ibis-project.org/docs/3.1.0/backends/DuckDB/#ibis.backends.duckdb.Backend.register--returns

**Comment by cpcloud** [comment]
> Nice feature, especially with the https URL support.
> 
> Maybe you should make it explicit in the docstring that this will read the full data in a in-memory (duckdb) table?

Ah, that's the thing though: it doesn't. It creates a view referencing the file.

> 
> If the file is large, it might be recommended instead to create an empty in-memory duckdb connection and subsequently register the parquet (or CSV) file as a foreign table:
> 
> [ibis-project.org/docs/3.1.0/backends/DuckDB/#ibis.backends.duckdb.Backend.register--returns](https://ibis-project.org/docs/3.1.0/backends/DuckDB/#ibis.backends.duckdb.Backend.register--returns)

This is exactly what it does :)

**Comment by ogrisel** [comment]
> This is exactly what it does :)

Nice! I think it might be worth making that explicit.

When the file is a remote https file, where does it go and when is it garbage collected?

**Comment by cpcloud** [comment]
> > This is exactly what it does :)
> 
> Nice! I think it might be worth making that explicit because the name `read_parquet` made me think it was read in memory ahead of query time.
> 
> When the file is a remote https file, where does it go and when is it garbage collected?
> 
> I mean, I assume it does not do HTTP range requests each time your do a query, or does it?

It does do range requests every time, because it's a view so it's not executed until you ask for it.

The tradeoffs here optimize for the case of fast construction of expressions, which in turn optimize for more rapid feedback. The cost is in repeated computation.

Hopefully, that's still useful if not optimal for some use cases :)

I'd be happy to call these `ibis.csv`/`ibis.parquet` or something else to make it less misleading about its in-memory status.

I did a bit of rudimentary data engineering in [this gist](https://gist.github.com/cpcloud/4676e93f505c9fb9071f6be7d1fc14c9) to determine the viability of the API and how it feels and so far I'm pretty happy with it.

**Comment by jcmkk3** [comment]
> I'd be happy to call these `ibis.csv`/`ibis.parquet` or something else to make it less misleading about its in-memory status.

I'm not sure if it is really the exact same behavior but polars uses [`scan_csv`](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.scan_csv.html) for a similar function.

Arrow has an [`open_csv`](https://arrow.apache.org/docs/python/generated/pyarrow.csv.open_csv.html#pyarrow.csv.open_csv) function, but again, I'm not sure if it is really analogous.

Both libraries also have an eager `read_csv` function.

Could `connect_csv` and `connect_parquet` be an option? They are a bit verbose, but would at least show up in autocomplete with the normal `connect`.

**Comment by ogrisel** [comment]
Maybe it's ok to keep the `read_*` names if the docstrings are explicit enough.

**Comment by cpcloud** [comment]
I've changed the name to `ibis.file`, which also allows users to avoid having to care about which function to call based on the file type they're accessing.

**Comment by jcrist** [comment]
I personally think "read" is the proper verb to use here, and is probably the word users coming from pandas/dask/pyspark will look for when browsing the API. Dask is also lazy and doesn't load any data, but we call things `read_csv`/`read_parquet`/.... anyway and no one has complained so far.

> I've changed the name to `ibis.file`, which also allows users to avoid having to care about which function to call based on the file type they're accessing.

I find the idea behind a single read-this-file command good, but:

- I worry about any file-type specific keywords making their way into this method, leading to hard to document/confusing kwargs. Having separate `read_csv`/`read_parquet` methods help avoid this issue IMO. We could keep a single API method that dispatched to these as well for users that want to auto-detect file types.
- I don't love "file", since it's a noun and not a verb. What's being done with the file? I'd be happier with `ibis.read`.
- Directories of csvs/parquet files are common, and something that many tools can read (and users may expect). Limiting a method from `some_string -> Table` to only handle "file" things feels limiting, and passing in a directory/glob string to a method called `file` would be confusing IMO.
- We're likely to get writer methods (#4444) at some point. Having parity in naming between the source and sink methods would be nice.

So in short:
- I'm happy with having a single method for now, but I'd prefer it was called `ibis.read`
- We might want to keep around `read_parquet`/`read_csv` for users wanting to avoid the auto-dispatch, and for a place to document filetype specific kwargs if those ever appear (the `ibis.read` method could then just say it "forwards kwargs to the file specific methods).

**Comment by cpcloud** [comment]
`read` seems fine too, at this point I don't really care about the name anymore as long as it's one of the few thrown out here. The functionality is really what matters :)

**Review by jreback — COMMENTED** [review]
could consider

ibis.read.parquet and similar

to avoid api lock-in

File: ibis/expr/api.py [hunk 1]
```diff
@@ -871,6 +873,33 @@ def row_number() -> ir.IntegerColumn:
...
+    ir.Table
+        Table expression representing a file
+    """
+    from ibis.config import _default_backend
+
+    con = _default_backend()
+    return con.register(str(path), **kwargs)
+
+
+read_csv = read_parquet = read
```
Comment (jcrist) [inline ibis/expr/api.py:900]: I don't love this, since the file type is still auto-determined from the file name (or scheme). If I have a file with a nonstandard extension I can't call `ibis.read_csv("test.data")` and have it work.

I'd prefer either:

- Dropping these for now
- Making them true top-level functions that don't autodetect based on the path/scheme.

**Review by jcrist — COMMENTED** [review]
Overall this looks pretty good to me, would love to have this functionality in ibis. Just one small comment, otherwise :+1:.

File: ibis/expr/api.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/api.py:900]: I'll drop them for now.

**Comment by cpcloud** [comment]
Confirmed that this works with globs!

### CSV Globs

```
❯ ls /data/nyc_taxi/yellow-subset
Permissions Size User  Date Modified Name
.rw-r--r--  352M cloud  5 Nov 08:58  yellow_tripdata_2021-10.csv
.rw-r--r--   66M cloud  5 Nov 08:58  yellow_tripdata_2021-11.csv.gz
.rw-r--r--   62M cloud  5 Nov 08:58  yellow_tripdata_2021-12.csv.gz
…/ibis on  read-file is 📦 v3.2.0 via 🐍 v3.10.8 via ❄️  impure (ibis310)
❯ ipython

In [1]: import ibis; ibis.options.interactive = True; from ibis import _

In [2]: t = ibis.read('csv:///data/nyc_taxi/yellow-subset/*')

In [3]: t
Out[3]:
┏━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━┓
┃ VendorID ┃ tpep_pickup_datetime ┃ tpep_dropoff_datetime ┃ passenger_count ┃ trip_distance ┃ RatecodeID ┃ store_and_fwd_flag ┃ PULocationID ┃ DOLocationID ┃ payment_type ┃ … ┃
┡━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━┩
│ int32    │ timestamp            │ timestamp             │ float64         │ float64       │ float64    │ string             │ int32        │ int32        │ int32        │ … │
├──────────┼──────────────────────┼───────────────────────┼─────────────────┼───────────────┼────────────┼────────────────────┼──────────────┼──────────────┼──────────────┼───┤
│        1 │ 2021-11-01 00:34:15  │ 2021-11-01 00:57:21   │             1.0 │          7.00 │        1.0 │ Y                  │          137 │          264 │            2 │ … │
│        1 │ 2021-11-01 00:00:29  │ 2021-11-01 00:14:13   │             1.0 │          3.20 │        1.0 │ N                  │          233 │          232 │            1 │ … │
│        1 │ 2021-11-01 00:27:51  │ 2021-11-01 00:34:21   │             2.0 │          1.80 │        1.0 │ N                  │           79 │          170 │            1 │ … │
│        1 │ 2021-11-01 00:01:56  │ 2021-11-01 00:25:18   │             0.0 │          4.70 │        1.0 │ N                  │          230 │           42 │            2 │ … │
│        1 │ 2021-11-01 00:10:44  │ 2021-11-01 00:29:06   │             1.0 │          3.50 │        1.0 │ N                  │          144 │           50 │            1 │ … │
│        1 │ 2021-11-01 00:36:21  │ 2021-11-01 00:54:46   │             2.0 │          3.40 │        1.0 │ N                  │          230 │          263 │            2 │ … │
│        1 │ 2021-11-01 00:12:05  │ 2021-11-01 00:31:06   │             0.0 │         14.30 │        1.0 │ N                  │          138 │          257 │            1 │ … │
│        1 │ 2021-11-01 00:49:43  │ 2021-11-01 00:59:38   │             1.0 │          1.40 │        1.0 │ N                  │          249 │          107 │            2 │ … │
│        2 │ 2021-11-01 00:10:05  │ 2021-11-01 00:37:59   │             1.0 │          6.29 │        1.0 │ N                  │          249 │           37 │            2 │ … │
│        2 │ 2021-11-01 00:26:13  │ 2021-11-01 00:37:37   │             1.0 │          2.15 │        1.0 │ N                  │           50 │           90 │            1 │ … │
│        2 │ 2021-11-01 00:41:44  │ 2021-11-01 00:46:21   │             1.0 │          0.73 │        1.0 │ N                  │           90 │          234 │            1 │ … │
│        … │ …                    │ …                     │               … │             … │          … │ …                  │            … │            … │            … │ … │
└──────────┴──────────────────────┴───────────────────────┴─────────────────┴───────────────┴────────────┴────────────────────┴──────────────┴──────────────┴──────────────┴───┘


In [4]: t.passenger_count.value_counts()
Out[4]:
┏━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ passenger_count ┃ count   ┃
┡━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ float64         │ int64   │
├─────────────────┼─────────┤
│             1.0 │ 7065734 │
│             2.0 │ 1545144 │
│             0.0 │  239513 │
│             3.0 │  407262 │
│             4.0 │  171978 │
│             5.0 │  220194 │
│             6.0 │  134720 │
│             7.0 │      29 │
│             9.0 │      14 │
│             8.0 │      11 │
│             nan │  366223 │
│               … │       … │
└─────────────────┴─────────┘
```

### Parquet Globs

```
In [5]: t = ibis.read('/data/ibis-testing/*.parquet')

In [6]: t
Out[6]:
┏━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃ id    ┃ bool_col ┃ tinyint_col ┃ smallint_col ┃ int_col ┃ bigint_col ┃ float_col ┃ double_col ┃ date_string_col      ┃ string_col           ┃ timestamp_col           ┃ year  ┃ month ┃
┡━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━┩
│ int32 │ boolean  │ int32       │ int32        │ int32   │ int64      │ float32   │ float64    │ binary               │ binary               │ timestamp               │ int32 │ int32 │
├───────┼──────────┼─────────────┼──────────────┼─────────┼────────────┼───────────┼────────────┼──────────────────────┼──────────────────────┼─────────────────────────┼───────┼───────┤
│  5160 │ True     │           0 │            0 │       0 │          0 │       0.0 │        0.0 │ b'06/01/10'          │ b'0'                 │ 2010-06-01 00:00:00.000 │  2010 │     6 │
│  5161 │ False    │           1 │            1 │       1 │         10 │       1.1 │       10.1 │ b'06/01/10'          │ b'1'                 │ 2010-06-01 00:01:00.000 │  2010 │     6 │
│  5162 │ True     │           2 │            2 │       2 │         20 │       2.2 │       20.2 │ b'06/01/10'          │ b'2'                 │ 2010-06-01 00:02:00.100 │  2010 │     6 │
│  5163 │ False    │           3 │            3 │       3 │         30 │       3.3 │       30.3 │ b'06/01/10'          │ b'3'                 │ 2010-06-01 00:03:00.300 │  2010 │     6 │
│  5164 │ True     │           4 │            4 │       4 │         40 │       4.4 │       40.4 │ b'06/01/10'          │ b'4'                 │ 2010-06-01 00:04:00.600 │  2010 │     6 │
│  5165 │ False    │           5 │            5 │       5 │         50 │       5.5 │       50.5 │ b'06/01/10'          │ b'5'                 │ 2010-06-01 00:05:00.100 │  2010 │     6 │
│  5166 │ True     │           6 │            6 │       6 │         60 │       6.6 │       60.6 │ b'06/01/10'          │ b'6'                 │ 2010-06-01 00:06:00.150 │  2010 │     6 │
│  5167 │ False    │           7 │            7 │       7 │         70 │       7.7 │       70.7 │ b'06/01/10'          │ b'7'                 │ 2010-06-01 00:07:00.210 │  2010 │     6 │
│  5168 │ True     │           8 │            8 │       8 │         80 │       8.8 │       80.8 │ b'06/01/10'          │ b'8'                 │ 2010-06-01 00:08:00.280 │  2010 │     6 │
│  5169 │ False    │           9 │            9 │       9 │         90 │       9.9 │       90.9 │ b'06/01/10'          │ b'9'                 │ 2010-06-01 00:09:00.360 │  2010 │     6 │
│  5170 │ True     │           0 │            0 │       0 │          0 │       0.0 │        0.0 │ b'06/02/10'          │ b'0'                 │ 2010-06-02 00:10:00.450 │  2010 │     6 │
│     … │ …        │           … │            … │       … │          … │         … │          … │ …                    │ …                    │ …                       │     … │     … │
└───────┴──────────┴─────────────┴──────────────┴─────────┴────────────┴───────────┴────────────┴──────────────────────┴──────────────────────┴─────────────────────────┴───────┴───────┘
```

**Comment by ogrisel** [comment]
This is great, however the error message could be improved to make explicit that a specific file format is not support, for instance with an excel file:

```python
>>> ibis.read("df.xslx")
Traceback (most recent call last):
  Input In [26] in <cell line: 1>
    ibis.read("df.xslx")
  File ~/code/ibis/ibis/expr/api.py:897 in read
    return con.register(str(path), **kwargs)
  File ~/code/ibis/ibis/backends/duckdb/__init__.py:264 in register
    sql, table_name, extensions_required = _generate_view_code(
  File ~/code/ibis/ibis/common/dispatch.py:103 in __call__
    return func(s, *args, **kwargs, **match.groupdict())
  File ~/code/ibis/ibis/backends/duckdb/__init__.py:105 in _file_or_url
    return _generate_view_code(f"{extension}://{path}", table_name=table_name, **kwargs)
  File ~/code/ibis/ibis/common/dispatch.py:103 in __call__
    return func(s, *args, **kwargs, **match.groupdict())
TypeError: ibis.backends.duckdb._file_or_url() got multiple values for keyword argument 'scheme'
```

or with mixed supported file formats matching in the same glob:

```python
>>> ls
df.csv      df.parquet
>>> ibis.read("df.*")
Traceback (most recent call last):
  Input In [32] in <cell line: 1>
    ibis.read("df.*")
  File ~/code/ibis/ibis/expr/api.py:897 in read
    return con.register(str(path), **kwargs)
  File ~/code/ibis/ibis/backends/duckdb/__init__.py:264 in register
    sql, table_name, extensions_required = _generate_view_code(
  File ~/code/ibis/ibis/common/dispatch.py:103 in __call__
    return func(s, *args, **kwargs, **match.groupdict())
  File ~/code/ibis/ibis/backends/duckdb/__init__.py:105 in _file_or_url
    return _generate_view_code(f"{extension}://{path}", table_name=table_name, **kwargs)
  File ~/code/ibis/ibis/common/dispatch.py:103 in __call__
    return func(s, *args, **kwargs, **match.groupdict())
TypeError: ibis.backends.duckdb._file_or_url() got multiple values for keyword argument 'scheme'
```

**Comment by cpcloud** [comment]
@ogrisel Ah, thanks for the feedback, great points. I'll see what I can do about making the error message less cryptic!

**Comment by cpcloud** [comment]
Alright, this now should have reasonable error messages for things that aren't supported.

File: ibis/backends/duckdb/__init__.py [hunk 2]
```diff
@@ -99,18 +117,16 @@ def _s3(full_path, table_name=None):
...
 
 
 @_generate_view_code.register(r".+", priority=1)
-def _default(_, **kwargs):
+def _default(path, **kwargs):
     raise ValueError(
-        """
-Unrecognized filetype or extension.
-Valid prefixes are parquet://, csv://, s3://, or file://
+        """Unrecognized file type or extension: {path}.
```
Comment (ogrisel) [inline ibis/backends/duckdb/__init__.py:126]: The `path` variable is not interpolated at the moment.


```suggestion
        f"""Unrecognized file type or extension: {path}.
```

**Review by ogrisel — COMMENTED** [review]
There are still some serious problems with the input validation logic:

```python
>>> import ibis
>>> ls
df.csv      df.parquet  df.xlsx
>>> ibis.read("csv://./*")
[1]    85705 segmentation fault  ipython
```

and:

```python
>>> import ibis
>>> ibis.read("parquet://./*")
Traceback (most recent call last):
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
    self.dialect.do_execute(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
    cursor.execute(statement, parameters)
  File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
    self.c.execute(statement, parameters)
InvalidInputException: Invalid Input Error: No magic bytes found at end of file '/private/tmp/tmpdata/df.xlsx'

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  Input In [3] in <cell line: 1>
    ibis.read("parquet://./*")
  File ~/code/ibis/ibis/expr/api.py:897 in read
    return con.register(str(path), **kwargs)
  File ~/code/ibis/ibis/backends/duckdb/__init__.py:256 in register
    self.con.execute(sql)
  File <string>:2 in execute
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py:402 in warned
    return fn(*args, **kwargs)
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3257 in execute
    return connection.execute(statement, *multiparams, **params)
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1365 in execute
    return self._exec_driver_sql(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1669 in _exec_driver_sql
    ret = self._execute_context(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1943 in _execute_context
    self._handle_dbapi_exception(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2124 in _handle_dbapi_exception
    util.raise_(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208 in raise_
    raise exception
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
    self.dialect.do_execute(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
    cursor.execute(statement, parameters)
  File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
    self.c.execute(statement, parameters)
ProgrammingError: (duckdb.InvalidInputException) Invalid Input Error: No magic bytes found at end of file '/private/tmp/tmpdata/df.xlsx'
[SQL: CREATE OR REPLACE VIEW "/private/tmp/tmpdata/*" AS
SELECT * FROM read_parquet('/private/tmp/tmpdata/*')]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

the segfault is probably revealing a lower level bug in the duckdb Python API which should probably be reported upstream with minimal reproducer.

**Comment by cpcloud** [comment]
> There are still some serious problems with the input validation logic:
> 
> ```python
> >>> import ibis
> >>> ls
> df.csv      df.parquet  df.xlsx
> >>> ibis.read("csv://./*")
> [1]    85705 segmentation fault  ipython
> ```
> 
> and:
> 
> ```python
> >>> import ibis
> >>> ibis.read("parquet://./*")
> Traceback (most recent call last):
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
>     self.dialect.do_execute(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
>     cursor.execute(statement, parameters)
>   File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
>     self.c.execute(statement, parameters)
> InvalidInputException: Invalid Input Error: No magic bytes found at end of file '/private/tmp/tmpdata/df.xlsx'
> 
> The above exception was the direct cause of the following exception:
> Traceback (most recent call last):
>   Input In [3] in <cell line: 1>
>     ibis.read("parquet://./*")
>   File ~/code/ibis/ibis/expr/api.py:897 in read
>     return con.register(str(path), **kwargs)
>   File ~/code/ibis/ibis/backends/duckdb/__init__.py:256 in register
>     self.con.execute(sql)
>   File <string>:2 in execute
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py:402 in warned
>     return fn(*args, **kwargs)
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3257 in execute
>     return connection.execute(statement, *multiparams, **params)
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1365 in execute
>     return self._exec_driver_sql(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1669 in _exec_driver_sql
>     ret = self._execute_context(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1943 in _execute_context
>     self._handle_dbapi_exception(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2124 in _handle_dbapi_exception
>     util.raise_(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208 in raise_
>     raise exception
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
>     self.dialect.do_execute(
>   File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
>     cursor.execute(statement, parameters)
>   File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
>     self.c.execute(statement, parameters)
> ProgrammingError: (duckdb.InvalidInputException) Invalid Input Error: No magic bytes found at end of file '/private/tmp/tmpdata/df.xlsx'
> [SQL: CREATE OR REPLACE VIEW "/private/tmp/tmpdata/*" AS
> SELECT * FROM read_parquet('/private/tmp/tmpdata/*')]
> (Background on this error at: https://sqlalche.me/e/14/f405)
> ```
> 
> the segfault is probably revealing a lower level bug in the duckdb Python API which should probably be reported upstream with minimal reproducer.

Oh boy 😬

**Comment by cpcloud** [comment]
@ogrisel Can you show the contents of those `df.*` files? I can't reproduce the segfault on my machine using a random collection of mixed files.

**Comment by ogrisel** [comment]
This is the same pandas dataframe exported 3 times:

```python
>>> pd.read_excel("df.xlsx")
    Unnamed: 0         a
0            0 -1.937078
1            1  0.528488
2            2  0.605134
3            3 -0.874927
4            4 -0.446203
..         ...       ...
95          95  0.207459
96          96 -2.141455
97          97  0.241468
98          98  0.411752
99          99 -0.940849

[100 rows x 2 columns]
```

I think they were generated with something like the following:

```python
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame({"a": np.random.randn(100)})
>>> df.to_excel("df.xlsx")
>>> df.to_parquet("df.parquet")
>>> df.to_csv("df.csv")
```

**Comment by ogrisel** [comment]
Not that the segfault can also happen with loading the parquet file with the CSV reader:

```python
In [4]: ibis.read("csv://./*.parquet").execute()
[1]    87118 segmentation fault  ipython
```

**Comment by cpcloud** [comment]
> Not that the segfault can also happen with loading the parquet file with the CSV reader:
> 
> ```python
> In [4]: ibis.read("csv://./*.parquet").execute()
> [1]    87118 segmentation fault  ipython
> ```

I think it _may_ be something specific to the way the file is being written, because I can't reproduce this when using files generated by duckdb:

```
In [4]: import ibis; ibis.options.interactive = True; from ibis import _

In [5]: t = ibis.read("csv:///data/ml-parquet/*.parquet")
ProgrammingError: (duckdb.InvalidInputException) Invalid Input Error: Error on line 1025: expected 5 values per row, but got 1. (DELIMITER='    ' (auto detected), QUOTE='"' (auto detected), ESCAPE='' (auto detected), HEADER=0 (auto detected), SAMPLE_SIZE=10240, IGNORE_ERRORS=0, ALL_VARCHAR=0)
[SQL: CREATE OR REPLACE VIEW "/data/ml-parquet/*_parquet" AS
SELECT * FROM read_csv('/data/ml-parquet/*.parquet', auto_detect=True)]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

**Comment by cpcloud** [comment]
This is reproducible outside of ibis entirely, so I don't think it should block this PR.

**Comment by cpcloud** [comment]
On duckdb latest `master` I cannot reproduce the segfault.

**Comment by ogrisel** [comment]
The code LGTM then. I think this should be used in examples and tutorials where appropriate. Maybe the example databases could be published as `.parquet` files (in additiona to sqlite db files). But this can be done in follow-up PRs.

**Review by gforsyth — APPROVED** [review]
This looks good to me, ~pending a rebase.~ scratch that, I think this is good to go in now.

**Comment by gforsyth** [comment]
Thanks @cpcloud !

**Comment by cpcloud** [comment]
Heck yeah!
